### DISCUSSION TITLE
Allowable subclasses for a function that accepts base class with a custom metaclass

### DISCUSSION THREAD

**Discussion by BaconPancakes** [microsoft/pyright#9082] [Q&A] [answered]
I'm using a library that uses Pydantic.

Pydantic's `BaseModel` has a custom metaclass. The library has a function that takes `BaseModel` as an argument. When I subclass `BaseModel` into `MySubclass`, this subclass is not assignable into that function that accepts `BaseModel.`

Is this intended behavior?

Example that can demonstrate this without Pydantic at all:

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.12&strict=true&code=GYJw9gtgBAhgRgYygSwgBzCALlAggIQGEBZAUyxgFgAoGmhAGxgGdmpiBPMigCgJPIwAlAC4aUCVAB0MutUYs2nfC1KEGzHhEELWAXk7dhY6pOmza8pq3YcAygFdE1zctXrmo8ZJlS5AE1JgKAgOAH1gBwA7HmZIUjDGZhFbFWY1DS9THwsaUIjonk5HZ0UhGgBiPBAAcwdtKJwwYKwONFIoACJW9oBtYqddZgBdTqgEGCiosBw4DsVkGqjSfygsMCg0GBAYbSxSEC647USNMea1to7Ot3SPMeQoqEiohCxkMCfO-JfOysluld%2BlxBKMUGxpjgFkt4AwOusuj1SMC0hkRp0gA)

```python
from abc import ABCMeta


class MyMeta(ABCMeta):
    ...

class MyBaseCls(metaclass=MyMeta):
    ...

class MySubclass(MyBaseCls):
    ...

def my_fun(some_cls: MyBaseCls):
    ...

my_fun(MySubclass)
# Argument of type "type[MySubclass]" cannot be assigned to parameter "some_cls" of type "MyBaseCls" in function "my_fun"
#   "type[MyMeta]" is not assignable to "type[MyBaseCls]"
```

**Comment by erictraut** [comment, marked as answer]
The annotation you're using for the `some_cls` parameter indicates that the corresponding argument value must be an instance of `MyBaseCls`. I think what you intend here is that the value must be a *class object* that is a subclass of `MyBaseCls`. If my assumption is correct, then you should change the annotation to `type[MyBaseCls]`. This eliminates the error.

Alternatively, if you want to accept any instance of the `MyMeta` metaclass, you could change the parameter's annotation to `MyMeta`.

**Reply by BaconPancakes** [reply]
Thanks - I missed that and you're absolutely correct. I wrongly assumed the library I am calling is type annotated correctly.
